openfe/protocols/openmm_utils/system_validation.py

```diff
@@ -0,0 +1,177 @@
+    ------
+    ValueError
+      * If there are multiple SolventComponents in the ChemicalSystem.
+      * If there is a SolventComponent and the `nonbonded_method` is
+        `nocutoff`.
+      * If the SolventComponent solvent is not water.
+    """
+    solv = [comp for comp in state.values()
+            if isinstance(comp, SolventComponent)]
+
+    if len(solv) > 0 and nonbonded_method.lower() == "nocutoff":
+        errmsg = "nocutoff cannot be used for solvent transformations"
+        raise ValueError(errmsg)
+
+    if len(solv) == 0 and nonbonded_method.lower() == 'pme':
+        errmsg = "PME cannot be used for vacuum transform"
+        raise ValueError(errmsg)
+
+    if len(solv) > 1:
+        errmsg = "Multiple SolventComponent found, only one is supported"
+        raise ValueError(errmsg)
+
+    if len(solv) > 0 and solv[0].smiles != 'O':
+        errmsg = "Non water solvent is not currently supported"
+        raise ValueError(errmsg)
+
+
+def validate_protein(state: ChemicalSystem):
+    """
+    Checks that the ChemicalSystem's ProteinComponent are suitable for the
+    alchemical protocol.
+
+    Parameters
+    ----------
+    state : ChemicalSystem
+      The chemical system to inspect.
+
+    Raises
+    ------
+    ValueError
+      If there are multiple ProteinComponent in the ChemicalSystem.
+    """
+    nprot = sum(1 for comp in state.values()
+                if isinstance(comp, ProteinComponent))
+
+    if nprot > 1:
+        errmsg = "Multiple ProteinComponent found, only one is supported"
+        raise ValueError(errmsg)
+
+
+ParseCompRet = Tuple[
+    Optional[List[SolventComponent]], Optional[List[ProteinComponent]],
+    Dict[SmallMoleculeComponent, OFFMol],
+]
+
+
+def get_components(state: ChemicalSystem) -> ParseCompRet:
+    """
+    Establish all necessary Components for the transformation.
+
+    Parameters
+    ----------
+    state : ChemicalSystem
+      ChemicalSystem to get all necessary components from.
+
+    Returns
+    -------
+    solvent_comp : Optional[SolventComponent]
+      If it exists, the SolventComponent for the state, otherwise None.
+    protein_comp : Optional[ProteinComponent]
+      If it exists, the ProteinComponent for the state, otherwise None.
+    openff_mols : Dict[str, openff.toolkit.Molecule]
+    """
+    def _get_single_comps(comp_list, comptype):
+        ret_comps = [comp for comp in comp_list
+                     if isinstance(comp, comptype)]
+        if ret_comps:
+            return ret_comps[0]
+        else:
+            return None
+
+    solvent_comp: Optional[SolventComponent] = _get_single_comps(
+        list(state.values()), SolventComponent
+    )
+
+    protein_comp: Optional[ProteinComponent] = _get_single_comps(
+        list(state.values()), ProteinComponent
+    )
+
+    off_small_mols = {}
+    for comp in state.components.values():
+        if isinstance(comp, SmallMoleculeComponent):
+            off_small_mols[comp] = comp.to_openff()
```

Comment:
why are we caching a `to_openff()` call here?  It would probably be easier to just return a list of `SmallMoleculeComponent` here and do `to_openff()` on demand rather than carry this dict around

Reply:
Because I have to call `to_openff()` about 5-6 other times further down, and it's probably much more expensive to do the roundtrip than to store a bit extra in memory.